src/imitation/data/types.py

```diff
@@ -69,26 +69,53 @@ def __post_init__(self):
 
 
 @dataclasses.dataclass(frozen=True)
-class Transitions:
-    """A batch of obs-act-obs-done transitions.
+class TransitionsMinimal:
+    """A batch of obs-act transitions.
 
-    Often generated by combining and processing several TrajectoryNoRew objects via
-    `flatten_trajectories()`.
+    This class and its subclasses are usually instantiated via
+    `imitation.data.rollout.flatten_trajectories`.
     """
 
     obs: np.ndarray
     """
     Previous observations. Shape: (batch_size, ) + observation_shape.
 
     The i'th observation `obs[i]` in this array is the observation seen
-    by the agent when choosing action `acts[i]`.
-
-    Invariant: obs.dtype == next_obs.dtype.
+    by the agent when choosing action `acts[i]`. `obs[i]` is not required to
+    be from the timestep preceding `obs[i+1]`.
     """
 
     acts: np.ndarray
     """Actions. Shape: (batch_size,) + action_shape."""
 
+    infos: np.ndarray
+    """Optional array of info dicts. Shape: (batch_size,)."""
```

Comment:
If it is optional, then why is it not `Optional[np.ndarray]`?

Yes, this would require changing the `Dataset` implementations to handle optional elements in order to type check, but I think lying about the types is just going to shift the error to run time.

Reply:
I think maybe "Optional" was just not meant to be there, since the rest of the code seems to assume it's always present.

In that case, what is the repercussion of making this mandatory? E.g. does it significantly inflate memory consumption? I think probably not in RAM (but should be tested), but will inflate serialized size a lot, so should we have some logic to optionally drop it (at least if it's all empty dicts) in `load`/`save`?

Reply:
I forgot to update the docstring after changing my mind to use `infos: np.array` rather than `Optional[np.ndarray]`.

Reply:
I was assuming that we would continuing using our default save format of Trajectories (I don't think I've ever saved Transitions), so I didn't consider the disk space consequences.

I could do a quick RAM benchmark.

Reply:
Fair point, I guess it's OK if we just save trajectories. Although note that `TransitionsMinimal` is not convertable into `Trajectory`.

Would be interested in the RAM repercussion here. I think an empty object is surprisingly large (compared to e.g. an int).

Reply:
Saving list of empty dicts in various formats:

https://gist.github.com/shwang/a925b013b9f813c319f08e2fec344bd2
```
+------------------+---------+-----------------+
|   save_format    | n_dicts | disk usage (KB) |
+------------------+---------+-----------------+
|      pickle      |    0    |        0        |
|      pickle      |   300   |        0        |
|      pickle      |   700   |        0        |
|      pickle      |   1000  |        5        |
|      pickle      | 1000000 |       5860      |
|      pickle      | 5000000 |      29305      |
|      numpy       |    0    |        0        |
|      numpy       |   300   |        1        |
|      numpy       |   700   |        3        |
|      numpy       |   1000  |        5        |
|      numpy       | 1000000 |       5861      |
|      numpy       | 5000000 |      29306      |
| numpy_compressed |    0    |        0        |
| numpy_compressed |   300   |        0        |
| numpy_compressed |   700   |        1        |
| numpy_compressed |   1000  |        2        |
| numpy_compressed | 1000000 |       1483      |
| numpy_compressed | 5000000 |       7413      |
+------------------+---------+-----------------+
```

Reply:
I'm confused by this:

```
5000000/(29305*1024)
0.16662045725985328
```

Does an empty dict really take up less than a byte?

Reply:
The number you calculated is the inverse quantity, the number of dicts per byte.

```
+------------------+---------+-----------------+--------------------+
|   save_format    | n_dicts | disk usage (KB) |   bytes per dict   |
+------------------+---------+-----------------+--------------------+
|      pickle      |    0    |        0        |        nan         |
|      pickle      |   300   |        1        | 3.4766666666666666 |
|      pickle      |   700   |        3        | 4.918571428571428  |
|      pickle      |   1000  |        5        |       5.243        |
|      pickle      | 1000000 |       5860      |      6.001241      |
|      pickle      | 5000000 |      29305      |     6.0018482      |
|      numpy       |    0    |        0        |        nan         |
|      numpy       |   300   |        1        | 4.986666666666666  |
|      numpy       |   700   |        3        | 5.565714285714286  |
|      numpy       |   1000  |        5        |       5.696        |
|      numpy       | 1000000 |       5861      |      6.001696      |
|      numpy       | 5000000 |      29306      |     6.0019392      |
| numpy_compressed |    0    |        0        |        nan         |
| numpy_compressed |   300   |        0        | 3.0633333333333335 |
| numpy_compressed |   700   |        1        | 2.2885714285714287 |
| numpy_compressed |   1000  |        2        |       2.106        |
| numpy_compressed | 1000000 |       1483      |      1.518984      |
| numpy_compressed | 5000000 |       7413      |     1.5183624      |
+------------------+---------+-----------------+--------------------+

```

Reply:
Oops, thanks for catching that, yes those results make more sense :) I find it weird the overhead increases superlinearly with the number of dicts, although it seems to asymptote at 6 bytes at least? I guess that's OK.

Is it reasonably fast, too? I seem to remember having some performance problems when pickling large numbers of Python objects, but they may have fixed this in recent versions.

Reply:
Oh that's interesting. numpy loading is way faster than pickle loading, even when compressed.

https://gist.github.com/shwang/ae6c96170548929a21abbab939dc3ae4
```
 % python scrap2.py
+------------------+---------+-----------------+--------------------+------------------+------------------+
|   save_format    | n_dicts | disk usage (KB) |    bytes/n_dict    |  serialize_time  | deserialize_time |
+------------------+---------+-----------------+--------------------+------------------+------------------+
|      pickle      |    0    |        0        |        nan         | [7.51018524e-05] | [9.10758972e-05] |
|      pickle      |   300   |        0        |        0.0         | [7.00950623e-05] |   [0.00020194]   |
|      pickle      |   700   |        0        |        0.0         |   [0.00016189]   |    [0.000247]    |
|      pickle      |   1000  |        5        |       5.243        |   [0.00018287]   |   [0.00026011]   |
|      pickle      | 1000000 |       5860      |      6.001241      |   [0.44883609]   |   [0.41594386]   |
|      pickle      | 5000000 |      29305      |     6.0018482      |   [2.64626718]   |   [3.25023007]   |
|      numpy       |    0    |        0        |        nan         |   [0.0172019]    |   [0.00144839]   |
|      numpy       |   300   |        1        | 4.986666666666666  |   [0.00096893]   |   [0.00035596]   |
|      numpy       |   700   |        3        | 5.565714285714286  |   [0.00078297]   |   [0.00015378]   |
|      numpy       |   1000  |        5        |       5.696        |   [0.00078106]   |   [0.00014973]   |
|      numpy       | 1000000 |       5861      |      6.001696      |   [0.69544315]   |   [0.00032496]   |
|      numpy       | 5000000 |      29306      |     6.0019392      |   [3.1576438]    |   [0.00021815]   |
| numpy_compressed |    0    |        0        |        nan         |   [0.00091314]   |   [0.00016975]   |
| numpy_compressed |   300   |        0        | 3.0633333333333335 |   [0.00230289]   |   [0.00061703]   |
| numpy_compressed |   700   |        1        | 2.2885714285714287 |   [0.00221896]   |   [0.00024486]   |
| numpy_compressed |   1000  |        2        |       2.106        |   [0.00181007]   |   [0.00020194]   |
| numpy_compressed | 1000000 |       1483      |      1.518984      |   [1.39837718]   |   [0.00025392]   |
| numpy_compressed | 5000000 |       7413      |     1.5183624      |   [7.0082581]    |   [0.00024199]   |
+------------------+---------+-----------------+--------------------+------------------+------------------+
```

Reply:
Wow that's quite dramatic.

I'm a bit confused why the deserialize time for NumPy is almost *constant* -- surely we should start seeing some variable costs show up with `5000000` dicts? I assume costs are more linear if we stored e.g. a random array of numbers?

It's possible NumPy has some very efficient representation for repeated references to the same object. But it's also possible it has some hack that's screwing up the benchmark (e.g. notice if you're loading in something that was just saved?)

Reply:
Caching seems really likely. I'll try moving the file first before loading

Reply:
Ok, making a lot more sense with caching. But still much much faster than pure pickle. Tried cloudpickle too, which is much slower than pickle.
https://gist.github.com/shwang/ae6c96170548929a21abbab939dc3ae4/edit
```
 % python scrap2.py
+------------------+---------+-----------------+--------------------+------------------+------------------+
|   save_format    | n_dicts | disk usage (KB) |    bytes/n_dict    |  serialize_time  | deserialize_time |
+------------------+---------+-----------------+--------------------+------------------+------------------+
|      pickle      |    0    |        0        |        nan         | [7.89165497e-05] | [9.08374786e-05] |
|      pickle      |   300   |        1        | 3.4766666666666666 | [8.48770142e-05] |   [0.00016999]   |
|      pickle      |   700   |        3        | 4.918571428571428  |   [0.00016689]   |   [0.00022817]   |
|      pickle      |   1000  |        5        |       5.243        |   [0.00022912]   |   [0.00041413]   |
|      pickle      | 1000000 |       5860      |      6.001241      |   [0.40788722]   |   [0.36054015]   |
|      pickle      | 5000000 |      29305      |     6.0018482      |   [2.32574725]   |    [3.259233]    |
|   cloudpickle    |    0    |        0        |        nan         |   [0.00019193]   | [7.82012939e-05] |
|   cloudpickle    |   300   |        0        | 2.0533333333333332 |   [0.00175905]   |   [0.00012112]   |
|   cloudpickle    |   700   |        1        | 2.0228571428571427 |   [0.00449204]   |   [0.00019622]   |
|   cloudpickle    |   1000  |        1        |       2.016        |   [0.00549197]   |   [0.00018096]   |
|   cloudpickle    | 1000000 |       1955      |      2.002284      |   [6.52187085]   |   [0.31458831]   |
|   cloudpickle    | 5000000 |       9776      |     2.0022764      |  [35.75733614]   |   [2.05795884]   |
|      numpy       |    0    |        0        |        nan         |   [0.03340602]   |   [0.00164199]   |
|      numpy       |   300   |        1        | 4.986666666666666  |   [0.00078702]   |   [0.00018001]   |
|      numpy       |   700   |        3        | 5.565714285714286  |     [0.0007]     |   [0.00017262]   |
|      numpy       |   1000  |        5        |       5.696        |   [0.00093722]   |   [0.00018883]   |
|      numpy       | 1000000 |       5861      |      6.001696      |   [0.52698088]   |   [0.00025892]   |
|      numpy       | 5000000 |      29306      |     6.0019392      |   [2.89590693]   |   [0.00024486]   |
| numpy_compressed |    0    |        0        |        nan         |   [0.0009737]    |   [0.00021696]   |
| numpy_compressed |   300   |        0        | 3.0633333333333335 |   [0.00159812]   |   [0.00035095]   |
| numpy_compressed |   700   |        1        | 2.2885714285714287 |   [0.00205636]   |   [0.00030589]   |
| numpy_compressed |   1000  |        2        |       2.106        |   [0.00208092]   |   [0.00024891]   |
| numpy_compressed | 1000000 |       1483      |      1.518984      |   [1.37592196]   |   [0.00024319]   |
| numpy_compressed | 5000000 |       7413      |     1.5183624      |   [7.09819221]   |   [0.00024414]   |
+------------------+---------+-----------------+--------------------+------------------+------------------+

```

Reply:
Another interesting thing I noticed is that `np.load` has a `mmap_mode` param which can be super useful for huge arrays.

Reply:
Oh wait, deserialize time is still the same. I was looking at the wrong column.